File: mqtt/native/mqtt-client.py
#
# Example MQTT client using paho-mqtt
#

import os
import time
import cbor
import binascii

import paho.mqtt.client as paho


# topic we publish and subscribe to
TOPIC = u'mqtt/test/mytopic1'

pid = os.getpid()
print('MQTT client starting with PID {}..'.format(pid))

client = paho.Client()

# called when the client has connected to the MQTT broker (or when the client reconnects
# upon a lost connection)
def on_connect(client, userdata, flags, rc):
    print('on_connect({}, {}, {}, {})'.format(client, userdata, flags, rc))

    # subscribe to a topic, this will fire on_message() upon receiving events
    client.subscribe(TOPIC, qos=0)

# called when a message has been received on a topic that the client subscribes to.
def on_message(client, userdata, msg):
    obj = cbor.loads(msg.payload)
    pid, seq, ran = obj[u'args']
    print('event received on topic {}: pid={}, seq={}, ran={}'.format(msg.topic, pid, seq, binascii.b2a_hex(ran)))

client.on_connect = on_connect
client.on_message = on_message

# connect to the universal transport we configured in Crossbar.io
client.connect('localhost', port=8080)

# connect to the dedicated MQTT transport we configured in Crossbar.io
# client.connect('localhost', port=1883)

# start the network loop in a background thread
# see: https://pypi.python.org/pypi/paho-mqtt#network-loop
client.loop_start()

# publish once a second forever ..
seq = 1
while True:
    obj = {
        u'args': [pid, seq, os.urandom(8)],
        u'eligible_authrole': [u'anonymous']
    }
    payload = cbor.dumps(obj)

    result, mid = client.publish(TOPIC, payload, qos=0, retain=True)
    print('event published: result={}, mid={}'.format(result, mid))

    seq += 1
    time.sleep(1)

File: mqtt/native/mqtt-client.js
//
// Example MQTT client using MQTT.js
//

var mqtt = require('mqtt');
var cbor = require('cbor');
var crypto = require('crypto');


// topic we publish and subscribe to
var topic = 'mqtt/test/mytopic1';


// payload decoder for receiving
function decode (payload) {
    //var obj = cbor.decodeFirstSync(payload);
    var obj = cbor.decodeAllSync(payload)[0];
    return obj;
}


// payload encoder for sending
function encode (pid, seq) {
    var obj = {
        args: [pid, seq, crypto.randomBytes(8)]
    };
    var payload = cbor.encode(obj);
    return payload;
}


// connect to the universal transport we configured in Crossbar.io
var client  = mqtt.connect('mqtt://localhost:8080');

// connect to the dedicated MQTT transport we configured in Crossbar.io
//var client  = mqtt.connect('mqtt://localhost:1883');

// called when the client has connected to the MQTT broker (or when the client reconnects
// upon a lost connection)
client.on('connect', function () {
    var pid = process.pid;
    var seq = 1;

    console.log('connect: pid=' + pid);

    client.subscribe(topic);

    function publish () {
        var payload = encode(pid, seq);
        client.publish(topic, payload);
        seq += 1;
    }

    setInterval(publish, 1000);
});


// called when a message has been received on a topic that the client subscribes to.
client.on('message', function (topic, payload) {
    // parse payload (a binary buffer) into a structured object
    var obj = decode(payload);
    console.log(obj);
    var pid = obj.args[0];
    var seq = obj.args[1];
    var ran = obj.args[2];
    console.log('event received topic ' + topic + ': pid=' + pid + ', seq=' + seq + ', ran=' + ran.toString('hex'));
    //client.end();
});
